espnet/distributed/pytorch_backend/launch.py

```diff
@@ -0,0 +1,167 @@
+#
+# SPDX-FileCopyrightText: Copyright (c) 2022 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+# SPDX-License-Identifier: Apache-2.0
+#
+# This code uses an official implementation of
+# distributed data parallel launcher as just a reference.
+# https://github.com/pytorch/pytorch/blob/v1.8.2/torch/distributed/launch.py
+#
+# One main difference is this code focuses on
+# launching simple function with given arguments.
+#
+
+
+import multiprocessing
+import os
+import signal
+import time
+
+
+_signalno_name_map = {
+    s.value: s.name for s in signal.valid_signals()
+    if isinstance(s, signal.Signals)
+}
+
+
+class WorkerError(multiprocessing.ProcessError):
+    def __init__(self, *, msg, exitcode, worker_id):
+        super(WorkerError, self).__init__(msg)
+        self._exitcode = exitcode
+        self._worker_id = worker_id
+
+    def __str__(self):
+        return (
+            f"worker[{self._worker_id}] failed with "
+            f"exitcode={self._exitcode}"
+        )
+
+    @property
+    def exitcode(self):
+        return self._exitcode
+
+    @property
+    def worker_id(self):
+        return self._worker_id
+
+
+class MainProcessError(multiprocessing.ProcessError):
+    def __init__(self, *, signal_no):
+        msg = (
+            f"{_signalno_name_map[signal_no]} received, "
+            f"exiting due to {signal.strsignal(signal_no)}."
+        )
+        super(MainProcessError, self).__init__(msg)
+        self._signal_no = signal_no
+        self._msg = msg
+
+    def __str__(self):
+        return self._msg
+
+    @property
+    def signal_no(self):
+        return self._signal_no
+
+
+def set_start_method(method):
+    assert method in ("fork", "spawn", "forkserver")
+    return multiprocessing.set_start_method(method)
+
+
+def _kill_processes(processes):
+    # TODO: This implementation can't stop all processes which have
+    # grandchildren processes launched within each child process
+    # directly forked from this script.
+    # Need improvement for more safe termination.
+    for p in processes:
+        try:
+            # NOTE: multiprocessing.Process.kill() was introduced in 3.7.
+            # https://docs.python.org/3.7/library/multiprocessing.html#multiprocessing.Process.kill
+            if not hasattr(p, "kill"):
+                p.terminate()
+            else:
+                p.kill()
+        except:  # noqa: E722
+            # NOTE: Ignore any exception happens during killing a process
+            # because this intends to send kill signal to *all* processes.
+            pass
+
+
+def launch(func, args, nprocs, master_addr="localhost", master_port=29500):
+    """
```

Comment:
Is the port number 29500 a hard coding? 

In espnet2, we select a free port number automatically.

https://github.com/espnet/espnet/blob/43ce0c69fb32961235534b348700dc6c74ad5792/espnet2/train/distributed_utils.py#L187-L198

Reply:
@kamo-naoyuki Thank you for the reviewing. The reason why I hard coded a port number here is that the original pytorch distributed also uses a hard coded value, and I thought library should have a default value. But, as you mentioned, a mechanism which automatically finds a free port is better for users. Let me update my PR.

Reply:
@kamo-naoyuki Sorry for my late action. I have updated this hard coded value. Default value is changed to `None` and it will be filled with automatically detected port number at the beginning of this function if `master_port` is not specified. Could you please check again if this modification makes sense?

Reply:
LGTM, thanks